helpers/browser/driver.js

```diff
@@ -216,18 +198,25 @@ class ChromeProtocol {
     });
 
     return this.connect()
-      .then(_ => sendCommand('Page.enable'))
-      .then(_ => sendCommand('Tracing.start', tracingOpts));
+      .then(_ => this.sendCommand('Page.enable'))
+      .then(_ => this.sendCommand('Tracing.start', tracingOpts));
   }
 
   endTrace() {
-    return new Promise((resolve, reject) => {
-      // When all Tracing.dataCollected events have finished, this event fires
-      this.on('Tracing.tracingComplete', _ => resolve(this._traceEvents));
+    return this.connect().then(_ => {
+      return new Promise((resolve, reject) => {
+        // Limit trace retrieval execution time.
+        const traceTimeoutId = setTimeout(_ => {
+          reject(new Error('Trace retrieval timed out'));
```

Comment:
not sure if this should still throw, but probably fine for now.

When gather error handling is added, this could, e.g., pass back an empty or partial trace with a `debugString` rather than just exiting all of lighthouse